src/set_theory/surreal.lean

```diff
@@ -269,28 +439,200 @@ instance : has_lt surreal   := ⟨lt⟩
-{ le_total := by rintro ⟨⟨x, ox⟩⟩ ⟨⟨y, oy⟩⟩; classical; exact
-    or_iff_not_imp_left.2 (λ h, le_of_lt oy ox (pgame.not_le.1 h)),
+{ le_total     := by rintro ⟨⟨x, ox⟩⟩ ⟨⟨y, oy⟩⟩; classical;
+                     exact or_iff_not_imp_left.2 (λ h, le_of_lt oy ox (pgame.not_le.1 h)),
   decidable_le := classical.dec_rel _,
-  ..surreal.ordered_add_comm_group }
+  .. surreal.ordered_add_comm_group }
+
+lemma nonneg_of_nsmul_pos (m : ℕ) {x : surreal} (hx : 0 < x) : 0 ≤ m • x :=
+begin
+  induction m with m hm,
+  { simp }, -- squeeze_simp fails?
+  { rw [succ_nsmul x m],
+    apply add_nonneg (le_of_lt hx) hm }
+end
+
+lemma pos_of_nsmul_pos {m : ℕ} {x : surreal} (hm : 0 < m) (hx : 0 < x) : 0 < m • x :=
+begin
+    induction m with m hm,
+    { exfalso, exact nat.lt_asymm hm hm },
+    { rw [succ_nsmul x m],
+      apply lt_add_of_pos_of_le hx (nonneg_of_nsmul_pos _ hx) }
+end
+
+lemma lt_of_nsmul_pos_lt {m : ℕ} {x y : surreal} (hm : 0 < m) (hxy : x < y) : m • x < m • y :=
+begin
+  rw ← sub_pos at *,
+  have : m • y - m • x = m • (y - x),
+    by { have := (nsmul_add y (-x) m).symm, rwa [neg_nsmul x m] at this },
+  rw this,
+  apply pos_of_nsmul_pos hm hxy,
+end
+
+lemma lt_of_gsmul_pos_lt {m : ℤ} {x y : surreal} (hm : 0 < m) (hxy : x < y) : m • x < m • y :=
+begin
+  cases m with m m,
+  { simp at *,
+    exact lt_of_nsmul_pos_lt hm hxy },
+  { exfalso,
+    rwa [← int.neg_succ_not_pos m] }
+end
+
+lemma gmul_cancel_of_pos {m : ℤ} {x y : surreal} (hm : 0 < m) (hmxy : m • x  = m • y) : x = y :=
+begin
+    contrapose hmxy,
+    cases (@ne_iff_lt_or_gt _ _ x y).1 hmxy with hmxy' hmxy',
+    { apply _root_.ne_of_lt,
+      exact lt_of_gsmul_pos_lt hm hmxy' },
+    { apply ne_of_gt,
+      change y < x at hmxy',
+      exact lt_of_gsmul_pos_lt hm hmxy' }
+end
+
+-- TODO: move this to a better location
+lemma int.pow_right_injective {x : ℤ} (h : 2 ≤ x) : function.injective (λ (n : ℕ), x ^ n) :=
+λ n m hnm, begin
+  lift x to ℕ using (zero_le_two).trans h,
+  norm_cast at h hnm,
+  exact nat.pow_right_injective h hnm
+end
+
+/-- The surreal number `half`. -/
+def half : surreal := ⟦⟨pgame.half, pgame.numeric_half⟩⟧
+
+/-- Powers of the surreal number `half`. -/
+def pow_half (n : ℕ) : surreal := ⟦⟨pgame.pow_half n, pgame.numeric_pow_half⟩⟧
+
+@[simp] lemma pow_half_zero : pow_half 0 = 1 := rfl
+
+@[simp] lemma pow_half_one : pow_half 1 = half := rfl
+
+@[simp] theorem add_half_self_eq_one : half + half = 1 :=
+quotient.sound pgame.add_half_self_equiv_one
+
+lemma double_pow_half_succ_eq_pow_half (n : ℕ) : 2 • pow_half n.succ = pow_half n :=
+begin
+  rw two_nsmul,
+  apply quotient.sound,
+  apply pgame.add_pow_half_succ_self_eq_pow_half,
+end
+
+/-- Exponentiation map from natural numbers to powers of integers. -/
+def pow (a : ℤ) (n : ℕ) : submonoid.powers a := ⟨a ^ n, n, rfl⟩
+
+/-- Logarithms from powers of integers to natural numbers. -/
+def log {a : ℤ} (p : submonoid.powers a) : ℕ :=
+nat.find $ (submonoid.mem_powers_iff a p.val).1 p.prop
```

Comment:
This should definitely be generalized to general monoids, and also the lemmata that can be (so for example, `log_power_eq_self` could be done with an `injective` predicate)

Reply:
Will do!